pkg/providers/cloudstack/cloudstack.go

```diff
@@ -576,10 +576,29 @@ func (p *cloudstackProvider) needsNewKubeadmConfigTemplate(workerNodeGroupConfig
 	return true, nil
 }
 
+// AnyImmutableFieldChanged Used by EKS-A controller and CLI upgrade workflow to compare generated CSDC/CSMC's from
+// CAPC resources in fetcher.go with those already on the cluster when deciding whether or not to generate and apply
+// new CloudStackMachineTemplates
 func AnyImmutableFieldChanged(oldCsdc, newCsdc *v1alpha1.CloudStackDatacenterConfig, oldCsmc, newCsmc *v1alpha1.CloudStackMachineConfig) bool {
-	if !oldCsdc.Spec.Equal(&newCsdc.Spec) {
+	if len(newCsdc.Spec.AvailabilityZones) != len(oldCsdc.Spec.AvailabilityZones) {
 		return true
 	}
+	oAzsMap := map[string]v1alpha1.CloudStackAvailabilityZone{}
+	for _, oAz := range oldCsdc.Spec.AvailabilityZones {
+		oAzsMap[oAz.Name] = oAz
+	}
+	for _, sAz := range newCsdc.Spec.AvailabilityZones {
```

Comment:
```suggestion
	for _, newAz := range newCsdc.Spec.AvailabilityZones {
```
If oAz means old availability zone then nAz or newAz may be more appropriate.